### PR TITLE
backport: merge bitcoin#18574, #18653, #18691, #18724, #18594, #19991, #19133, #19643, #20002, #19354, partial #19998 (cli backports)

### PR DESCRIPTION
**Pull request by kittywhiskers**

### CODE REVIEW

**Review by PastaPastaPasta — DISMISSED** [review]
utACK for merging via merge commit

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -498,62 +554,46 @@ static int CommandLineRPC(int argc, char *argv[])
...
+            if (error.isObject()) {
+                const UniValue& errCode = find_value(error, "code");
+                const UniValue& errMsg = find_value(error, "message");
+                strPrint = errCode.isNull() ? "" : ("error code: " + errCode.getValStr() + "\n");
+
+                if (errMsg.isStr()) {
+                    strPrint += ("error message:\n" + errMsg.get_str());
+                }
+                if (errCode.isNum() && errCode.get_int() == RPC_WALLET_NOT_SPECIFIED) {
+                    strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
```
Comment (knst) [inline src/bitcoin-cli.cpp]: dash-cli

**Review by knst — CHANGES_REQUESTED** [review]
LGTM but see a comment about dashification

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (kittywhiskers) [reply src/bitcoin-cli.cpp]: Resolved in latest push!

File: doc/tor.md [hunk 2]
```diff
@@ -53,11 +53,12 @@ config file): *Needed for Tor version 0.2.7.0 and older versions of Tor only. Fo
...
-	HiddenServicePort 9999 127.0.0.1:9999
-	HiddenServicePort 19999 127.0.0.1:19999
+	HiddenServicePort 9999 127.0.0.1:9996
+	HiddenServicePort 19999 127.0.0.1:19996
 
-The directory can be different of course, but (both) port numbers should be equal to
-your dashd's P2P listen port (9999 by default).
+The directory can be different of course, but virtual port numbers should be equal to
+your dashd's P2P listen port (9999 by default), and target addresses and ports
+should be equal to binding address and port for inbound Tor connections (127.0.0.1:9999 by default).
```
Comment (UdjinM6) [inline doc/tor.md:61]: 19991: s/9999/9996/

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -299,6 +302,211 @@ class GetinfoRequestHandler: public BaseRequestHandler
...
+    UniValue ProcessReply(const UniValue& batch_in) override
+    {
+        const std::vector<UniValue> batch{JSONRPCProcessBatchReply(batch_in)};
+        if (!batch[ID_PEERINFO]["error"].isNull()) return batch[ID_PEERINFO];
+        if (!batch[ID_NETWORKINFO]["error"].isNull()) return batch[ID_NETWORKINFO];
+
+        const UniValue& networkinfo{batch[ID_NETWORKINFO]["result"]};
+        if (networkinfo["version"].get_int() < 209900) {
+            throw std::runtime_error("-netinfo requires bitcoind server to be running v0.21.0 and up");
+        }
```
Comment (UdjinM6) [inline src/bitcoin-cli.cpp:379]: 19643: needs dashification

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/interface_bitcoin_cli.py [hunk 4]
```diff
@@ -76,43 +79,61 @@ def run_test(self):
             assert_equal(self.nodes[0].cli.getwalletinfo(), wallet_info)
 
             # Setup to test -getinfo and -rpcwallet= with multiple wallets.
-            wallets = ['', 'Encrypted', 'secret']
-            amounts = [Decimal('59.999928'), Decimal(9), Decimal(31)]
+            wallets = [self.default_wallet_name, 'Encrypted', 'secret']
```
[the remark this answers is not in the document]
Comment (knst) [reply test/functional/interface_bitcoin_cli.py:94]: due to bitcoin#20034 that is already backported.

File: doc/tor.md (hunk 2 above)
Comment (kittywhiskers) [reply doc/tor.md]: Fixed in latest push

File: src/bitcoin-cli.cpp [hunk 5]
```diff
@@ -380,14 +380,12 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
 
     // check if we should use a special wallet endpoint
     std::string endpoint = "/";
-    if (!gArgs.GetArgs("-rpcwallet").empty()) {
-        std::string walletName = gArgs.GetArg("-rpcwallet", "");
-        char *encodedURI = evhttp_uriencode(walletName.data(), walletName.size(), false);
+    if (rpcwallet) {
```
Comment (knst) [inline src/bitcoin-cli.cpp:581]: I think you should also include release notes to this PR (this backport):
    Merge #19354: doc: add release note for -getinfo displaying multiwallet balances

Comment (kittywhiskers) [reply src/bitcoin-cli.cpp:581]: Resolved in latest push, good catch!

**Review by knst — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Pull request merged**